File: 09/taehyuk/1268.py
from typing import List

# https://www.geeksforgeeks.org/trie-insert-and-search/
class TrieNode:
    def __init__(self):
        self.child = [None] * 26
        self.word_end = False

        self.suggestions = []        

def insert_key(root: TrieNode, key: str):
    curr = root

    for c in key:
        index = ord(c) - ord('a')
        
        if curr.child[index] is None:
            curr.child[index] = TrieNode()
        
        curr = curr.child[index]
        
        if len(curr.suggestions) < 3:
            curr.suggestions.append(key)
            curr.suggestions.sort()
        else:
            if key < curr.suggestions[-1]:
                curr.suggestions[-1] = key
                curr.suggestions.sort()
                
    curr.word_end = True

def getSuggestions(root: TrieNode, prefix: str) -> List[List[str]]:
    curr = root
    result = []

    for c in prefix:
        index = ord(c) - ord('a')
        if curr.child[index] is None:
            break
        curr = curr.child[index]
        result.append(curr.suggestions)

    while len(result) < len(prefix):
        result.append([])

    return result

class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        
        root = TrieNode()
        for product in products:
            insert_key(root, product)
        
        return getSuggestions(root, searchWord)

File: 09/ahnjaewoo/1268.cpp
class Solution {
public:
	vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
		sort(products.begin(), products.end());
		vector<vector<string>> answer;
		vector<string> hubo = products;
		for (int i = 0; i < searchWord.size(); i++) {
			string prefix = searchWord.substr(0, i + 1);
			int matched = 0;
			vector<string> matchedProducts;
			for (int j = 0; j < hubo.size(); j++) {
				if (hubo[j].substr(0, i + 1) == prefix) {
					matched++;
					matchedProducts.push_back(hubo[j]);
				}
				if (matched == 3) break;
			}
			if (matched < 3) hubo = matchedProducts;
			answer.push_back(matchedProducts);
		}
		return answer;
	}
};